File: 393.utf-8-validation.cpp
/*
 * @lc app=leetcode id=393 lang=cpp
 *
 * [393] UTF-8 Validation
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
    int eigthBit = 1 << 7;

public:
    bool validUtf8(vector<int> &data)
    {
        int neededOneZeroCount = 0;
        for (auto val : data)
        {
            if (neededOneZeroCount)
            {
                if (!validOneZero(val))
                {
                    return false;
                }
                else
                {
                    --neededOneZeroCount;
                }
            }
            else if (val & eigthBit)
            {
                int num = numOfOne(val);
                if (val > 0)
                {
                    neededOneZeroCount += num - 1;
                }
                else
                {
                    return false;
                }
            }
        }
        return neededOneZeroCount == 0;
    }
    short numOfOne(const int &val)
    {
        short count = 0;
        int b = (1 << 7);
        while (val & b)
        {
            b >>= 1;
            ++count;
        }
        return count > 4 || count == 1 ? -1 : count;
    }
    bool validOneZero(const int &val)
    {
        return (val & (1 << 7)) && !(val & (1 << 6));
    }
};
// @lc code=end

File: 393.utf-8-validation.java
/*
 * @lc app=leetcode id=393 lang=java
 *
 * [393] UTF-8 Validation
 */

// @lc code=start
class Solution {
    public boolean validUtf8(int[] data) {
        byte neededOneZeroCount = 0;
        for (var val : data) {
            if (neededOneZeroCount == 0) {
                if ((val & (1 << 7)) == 0) {
                    continue;
                } else if ((val >> 5) == 0x06) {
                    neededOneZeroCount = 1;
                } else if ((val >> 4) == 0x0e) {
                    neededOneZeroCount = 2;
                } else if ((val >> 3) == 0x1e) {
                    neededOneZeroCount = 3;
                } else {
                    return false;
                }
            } else {
                if ((val >> 6) != 0x02) {
                    return false;
                } else {
                    --neededOneZeroCount;
                }
            }
        }
        return neededOneZeroCount == 0;
    }
}
// @lc code=end
